sdks/go/pkg/beam/io/synthetic/step.go

```diff
@@ -144,49 +143,130 @@ func (fn *sdfStepFn) Setup() {
+	return b
+}
+
+// InitialSplits is only applicable if Splittable is set to true, and determines
+// the number of initial splits to perform in the step's SplitRestriction
+// method. Restrictions in synthetic steps represent the number of elements to
+// emit for each input element, as defined by the OutputPerInput config field,
+// and this split is performed evenly across that number of elements.
+//
+// Each resulting restriction will have at least 1 element in it, and each
+// element being emitted will be contained in exactly one restriction. That
+// means that if the desired number of splits is greater than the OutputPerInput
+// N, then N initial restrictions will be created, each containing 1 element.
+//
+// Valid values are in the range of [1, ...] and the default value is 1. Values
+// of 0 (and below) are invalid as they would result in dropping elements that
+// are expected to be emitted.
+func (b *StepConfigBuilder) InitialSplits(val int) *StepConfigBuilder {
+	b.cfg.initialSplits = val
+	return b
+}
+
+// Build constructs the StepConfig initialized by this builder. It also performs
+// error checking on the fields, and panics if any have been set to invalid
+// values.
+func (b *StepConfigBuilder) Build() StepConfig {
+	if b.cfg.initialSplits <= 0 {
+		panic(fmt.Sprintf("StepConfig.InitialSplits must be >= 1. Got: %v", b.cfg.initialSplits))
+	}
+	if b.cfg.outputPerInput < 0 {
+		panic(fmt.Sprintf("StepConfig.OutputPerInput cannot be negative. Got: %v", b.cfg.outputPerInput))
 	}
+	return b.cfg
 }
 
 // StepConfig is a struct containing all the configuration options for a
-// synthetic step.
+// synthetic step. It should be created via a StepConfigBuilder.
 type StepConfig struct {
-	// OutputPerInput is the number of outputs to emit per input received. Each
-	// output is identical to the original input. A value of 0 drops each input.
-	OutputPerInput int
-
-	// FilterRatio indicates the random chance that an input will be filtered
-	// out, meaning that no outputs will get emitted for it. For example, a
-	// FilterRatio of 0.25 means that 25% of inputs will get filtered out.
-	FilterRatio float64
-
-	// Splittable indicates whether the step should use the splittable DoFn or
-	// non-splittable DoFn implementation. Splittable steps will split the
-	// number of OutputPerInput into restrictions, so it is most useful for
-	// steps with a high OutputPerInput.
-	Splittable bool
-
-	// InitialSplits is only applicable if Splittable is set to true, and
-	// determines the number of initial splits to perform in the step's
-	// SplitRestriction method. Note that in some edge cases, the number of
-	// splits performed might differ from this config value. Each restriction
-	// will always have one element in it, and at least one restriction will
-	// always be output, so the number of splits will be in the range of [1, N]
-	// where N is the size of the original restriction.
-	InitialSplits int
+	outputPerInput int
```

Comment:
Same comment here WRT the unexported fields being unserializable.

Reply:
Done.